packages/common/src/helper.ts

```diff
@@ -228,112 +230,251 @@ interface ModifyTemplateForDynamicTableArg {
-  const diffMap = new Map<number, number>();
-  const keys = Array.from(tmpDiffMap.keys()).sort((a, b) => a - b);
-  let additionalHeight = 0;
+  setKeyAndSchema(key: string, schema: Schema): void {
+    this.key = key;
+    this.schema = schema;
+  }
 
-  for (const key of keys) {
-    const value = tmpDiffMap.get(key) as number;
-    const newValue = value + additionalHeight;
-    diffMap.set(key + additionalHeight, newValue);
-    additionalHeight += newValue;
+  setWidth(width: number): void {
+    this.width = width;
   }
 
-  return diffMap;
-};
+  setHeight(height: number): void {
+    this.height = height;
+  }
 
-export const normalizePositionsAndPageBreak = (
-  template: Template,
-  diffMap: Map<number, number>
-): Template => {
-  if (!isBlankPdf(template.basePdf) || diffMap.size === 0) {
-    return template;
+  setPadding(padding: [number, number, number, number]): void {
+    this.padding = padding;
   }
 
-  const returnTemplate: Template = { schemas: [{}], basePdf: template.basePdf };
-  const pages = returnTemplate.schemas;
-  const pageHeight = template.basePdf.height;
-  const paddingTop = template.basePdf.padding[0];
-  const paddingBottom = template.basePdf.padding[2];
+  setPosition(position: { x: number; y: number }): void {
+    this.position = position;
+  }
 
-  for (let i = 0; i < template.schemas.length; i += 1) {
-    const schemaObj = template.schemas[i];
-    if (!pages[i]) pages[i] = {};
+  insertChild(child: Node): void {
+    const index = this.getChildCount();
+    child.setIndex(index);
+    this.children.splice(index, 0, child);
+  }
+
+  getChildCount(): number {
+    return this.children.length;
+  }
 
-    for (const [key, schema] of Object.entries(schemaObj)) {
-      const { position, height } = schema;
-      let newY = position.y;
-      let pageCursor = i;
+  getChild(index: number): Node {
+    return this.children[index];
+  }
+}
+
+function createPage(basePdf: BlankPdf) {
+  const page = new Node({ ...basePdf });
+  page.setPadding(basePdf.padding);
+  return page;
+}
+
+function createNode(arg: {
+  key: string;
+  schema: Schema;
+  position: { x: number; y: number };
+  width: number;
+  height: number;
+}) {
+  const { position, width, height, key, schema } = arg;
+  const node = new Node({ width, height });
+  node.setPosition(position);
+  node.setKeyAndSchema(key, schema);
+  return node;
+}
 
-      for (const [diffKey, diffValue] of diffMap) {
-        if (newY > diffKey) {
-          newY += diffValue;
+async function createOnePage(
+  arg: {
+    basePdf: BlankPdf;
+    schemaObj: Record<string, Schema>;
+  } & Omit<ModifyTemplateForDynamicTableArg, 'template'>
+): Promise<Node> {
+  const { basePdf, schemaObj, input, options, _cache, getDynamicHeights } = arg;
+  const page = createPage(basePdf);
+
+  const schemaEntries = Object.entries(schemaObj);
+  const schemaPositions: number[] = [];
+  const sortedSchemaEntries = schemaEntries.sort((a, b) => a[1].position.y - b[1].position.y);
```

Comment:
As you pointed out, this sorting affects the rendering order specified by the user by sorting based on position.y.

However, if we remove this sorting, when a schema (let's call it A) changes its height, we won't be able to shift downwards any schemas positioned below A if their order is set to be above A.

![CleanShot 2024-09-10 at 09 49 16](https://github.com/user-attachments/assets/399241e5-d7e4-4520-adb3-f5ff33e8dd73)

Reply:
With this modification, I've made it restore to the original order.
https://github.com/pdfme/pdfme/pull/589/commits/b276170b090c4a502f2ab6e30daf3fff0d29cc7d

Reply:
Tables are really hard 😄 

This is very clever. 

I can see an issue though if you _don't_ want to push content down due to the table.
See this example where I have modified the invoice design so that the bottom section is now more like a 'footer'. 

In this case we might want the table to use the available space without pushing items further down. However, you could also want some elements related to the table to be moved but not others 🤯 

https://github.com/user-attachments/assets/ebade097-9c2c-4793-9d6d-10b63fc9d228

In this case you could:
* group schemas for behaviour - fixed or shift
* set an attribute on individual schemas that would lock them if the table shifts.

However, this could be added in future. You have done an amazing job implementing this existing use case.